File: backend/internal/domain/analytics/analytics.go
// Package analytics is the reporting domain (docs/build/PHASE-6-ADMIN-OBS.md §2,
// FR-AN-001/002). It serves project analytics and the org usage dashboard from the
// nightly rollup (project_stats_daily) and the usage aggregates — never live
// aggregates over the operational tables. Stdlib-only.
package analytics

import "time"

// DailyStat is one project_stats_daily row (the Phase-5 rollup grain), the raw
// input the project-analytics view is assembled from.
type DailyStat struct {
	ProjectID       string
	Day             time.Time
	CreatedCount    int
	CompletedCount  int
	ColumnSnapshot  map[string]int // column_id → open task count
	AvgCycleSeconds *int64
}

// ProjectAnalytics is the assembled per-project view (FR-AN-001): the daily series
// plus roll-up totals over the requested window. CumulativeFlow is the per-day
// column snapshot series (for the CFD chart); the series is ordered by day ascending.
type ProjectAnalytics struct {
	ProjectID      string
	From           time.Time
	To             time.Time
	Series         []DailyStat
	TotalCreated   int
	TotalCompleted int
	// AvgCycleSeconds is the mean of the days that recorded one (nil ⇒ none did).
	AvgCycleSeconds *int64
}

// UsageMetric names a metered dimension mirrored from billing usage_records.
type UsageMetric string

const (
	MetricActiveMembers UsageMetric = "active_members"
	MetricStorageBytes  UsageMetric = "storage_bytes"
	MetricAPICalls      UsageMetric = "api_calls"
)

// UsageDashboard is the org usage view (FR-AN-002): the current period's metered
// dimensions plus an invoice estimate in minor units. Seats/Storage are the latest
// values; APICalls is the period sum.
type UsageDashboard struct {
	PeriodStart     time.Time
	PeriodEnd       time.Time
	Seats           int64
	StorageBytes    int64
	APICalls        int64
	EstimatedTotal  int64 // minor units (invariant #5)
	Plan            string
}

File: web/lib/api/analytics.ts
import { apiFetch } from './client';
import type { ProjectAnalytics, UsageDashboard } from './types';

// Analytics surface (FR-AN-001/002). Both read nightly rollups, never live
// aggregation. Project analytics = any member (read:org); org usage = ADMIN+
// (read:billing), so the usage caller must guard on role.

/** Org usage dashboard (seats/storage/api-calls + estimate). ADMIN+ only. */
export function getUsageDashboard(orgId: string): Promise<UsageDashboard> {
  return apiFetch(`/orgs/${orgId}/usage`);
}

/** A project's rollup analytics over an optional [from,to] day window (YYYY-MM-DD). */
export function getProjectAnalytics(
  orgId: string,
  projectId: string,
  window: { from?: string; to?: string } = {},
): Promise<ProjectAnalytics> {
  const qs = new URLSearchParams();
  if (window.from) qs.set('from', window.from);
  if (window.to) qs.set('to', window.to);
  const suffix = qs.toString() ? `?${qs.toString()}` : '';
  return apiFetch(`/orgs/${orgId}/projects/${projectId}/analytics${suffix}`);
}
